### DISCUSSION TITLE
Looking for help in recovering ext4.vhdx

### DISCUSSION THREAD

**Discussion by Zelaraus** [microsoft/WSL#8839] [General]
**Version**

Windows 11 Home 10.0.22621 Build 22621

**WSL Version**

WSL2

**Distro Version**

Ubuntu 20.04

I have mounted the solid state drive of one laptop to another and attached the ext4 virtual disk image on Disk Management. 

There is no option on right clicking the attached disk other than "New Simple Volume". (Shows unallocated drive in the box)

 I did that but it seemed it overwrote on the vhdx file such that its partition turned into NFTS.

Is there any way to recover the original partition for the ext4 system? I have tried multiple solutions / efforts available online but none of them works so far.

Thank you very much!

**Comment by Biswa96** [comment]
If you use the new WSL Preview there is a new command `wsl.exe --import-in-place` to take an existing .vhdx file and register it as a distro.

**Reply by cool-delete** [reply]
How to install WSL Preview in the win10 system

**Reply by Biswa96** [reply]
The easier way is from here https://apps.microsoft.com/store/detail/windows-subsystem-for-linux/9P9TQF7MRM4R?hl=en-us&gl=us

**Comment by NotTheDr01ds** [comment]
I'm assuming that you haven't written any files to the vhdx; just the format.  If that's the case, then maybe/hopefully there's a recovery option.  

[Accidentally reformatted a EXT4 drive to NTFS; lost all data. What are my options?](https://superuser.com/q/274615/1210833) and a [few others](https://www.google.com/search?q=recover+ext4+partition+that+was+formatted+ntfs+site:superuser.com).

The general consensus seems to be [TestDisk](https://www.cgsecurity.org/wiki/TestDisk), but also some mentions that it isn't necessarily an easy process -- Make sure to read the doc thoroughly.  And, of course, if you have space, create a copy of the `ext4.vhdx` before proceeding, of course.

If you need to TestDisk under WSL (I'm not sure if you will), then you'll need the WSL Preview (available from the Microsoft Store or here in "Releases"), since as @Biswa96 mentioned, that adds the ability to mount `.vhdx` directly.

Try:

* Starting your WSL distribution
* `lsblk`.  Take note of which devices are available
* Switch to PowerShell, and `wsl --mount --bare --vhd <file>`
* Switch back to the WSL distribution and `lsblk` again to see which device has been added.  This will be the drive, of course.

You should then be able to use TestDisk (the Linux version) on the drive/partition.

Note also that someone on Super User mentioned being able to [recover with a `fsck`](https://superuser.com/a/1204121/1210833), but they also mention that it was a "bit different" scenario.

**Comment by Zelaraus** [comment]
Many thanks for both of your replies @Biswa96 @NotTheDr01ds ! It seems the version installed for wsl is by default the preview... Luckily this feature may get across the Mount-VHD not available in the home version.

Before advancing to wsl --mount, lsblk yields 

NAME MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
sda    8:0    0 366.8M  1 disk
sdb    8:16   0     1G  0 disk [SWAP]
sdc    8:32   0     1T  0 disk /mnt/wslg/distro
                               /

After doing

wsl --mount --bare --vhd ext4.vhdx
The operation completed successfully.
,
 
it gives

NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
sda      8:0    0 366.8M  1 disk
sdb      8:16   0     1G  0 disk [SWAP]
sdc      8:32   0     1T  0 disk /mnt/wslg/distro
                                 /
sdd      8:48   0   256G  0 disk
├─sdd1   8:49   0    16M  0 part
└─sdd2   8:50   0   256G  0 part

**Comment by Zelaraus** [comment]
In the end there seems to be no luck on my side:

the testdisk detects GPT partition table at the beginning (similar to this https://forum.cgsecurity.org/phpBB3/viewtopic.php?t=6436), but might not be able to detect anything except the reformatted partition for the mounted drive.

Disk /dev/sdd - 274 GB / 256 GiB - CHS 33418 255 63
     Partition               Start        End    Size in sectors
>P MS Data                    32768  536866815  536834048 [New Volume]

Also tried fsck:

sudo fsck.ext4 -v /dev/sdd
[sudo] password for Zelaraus:
e2fsck 1.46.5 (30-Dec-2021)
ext2fs_open2: Bad magic number in super-block
fsck.ext4: Superblock invalid, trying backup blocks...
fsck.ext4: Bad magic number in super-block while trying to open /dev/sdd

The superblock could not be read or does not describe a valid ext2/ext3/ext4
filesystem.  If the device is valid and it really contains an ext2/ext3/ext4
filesystem (and not swap or ufs or something else), then the superblock
is corrupt, and you might try running e2fsck with an alternate superblock:
    e2fsck -b 8193 <device>
 or
    e2fsck -b 32768 <device>

Found a gpt partition table in /dev/sdd
